python/ray/serve/router.py

```diff
@@ -55,148 +57,158 @@ def __init__(
 
-        for new_replica_handle in added:
-            self.in_flight_queries[new_replica_handle] = set()
+            for new_replica_handle in added:
+                self.in_flight_queries[new_replica_handle] = set()
 
-        for removed_replica_handle in removed:
-            # Delete it directly because shutdown is processed by controller.
-            del self.in_flight_queries[removed_replica_handle]
+            for removed_replica_handle in removed:
+                # Delete it directly because shutdown is processed
+                # by controller.
+                del self.in_flight_queries[removed_replica_handle]
 
-        if len(added) > 0 or len(removed) > 0:
-            self.replica_iterator = itertools.cycle(
-                self.in_flight_queries.keys())
-            logger.debug(
-                f"ReplicaSet: +{len(added)}, -{len(removed)} replicas.")
-            self.config_updated_event.set()
+            if len(added) > 0 or len(removed) > 0:
+                self.replica_iterator = itertools.cycle(
+                    self.in_flight_queries.keys())
+                logger.debug(
+                    f"ReplicaSet: +{len(added)}, -{len(removed)} replicas.")
+                self._drain_pending_queue()
 
-    def _try_assign_replica(self, query: Query) -> Optional[ray.ObjectRef]:
-        """Try to assign query to a replica, return the object ref if succeeded
-        or return None if it can't assign this query to any replicas.
+    def _on_query_tracker_completed(
+            self,
+            _trakcer_result: Any,
+            *,
+            replica_handle: ActorHandle,
+            tracker_ref: ray.ObjectRef,
+    ):
+        """Callback after each query is completed."""
+        with self.lock:
+            if replica_handle in self.in_flight_queries:
+                self.in_flight_queries[replica_handle].remove(tracker_ref)
+                self._drain_pending_queue()
+
+    def _drain_pending_queue(self):
+        with self.lock:
+            num_pending = len(self.pending_queue)
+            if num_pending == 0:
+                # No pending queries.
+                return
+
+            num_possible = (
+                self.max_concurrent_queries * len(self.in_flight_queries) -
+                sum(map(len, self.in_flight_queries.values())))
+            if num_possible <= 0:
+                # All replicas busy.
+                return
+
+            if num_pending > num_possible:
+                # More pending queries than possible, clip the value.
+                num_to_send = num_pending - num_possible
+            else:
+                # We can send all the queries.
+                num_to_send = num_pending
+
+            logger.debug(f"Draining {num_to_send} queries")
+            for _ in range(num_to_send):
+                self._assign_replica(*self.pending_queue.pop(0))
+
+            self.num_queued_queries_gauge.set(len(self.pending_queue))
```

Comment:
This logic seems a little bit unnecessarily complicated. Can't we just do something like have a single loop where we iterate over the pending queries and try to assign them to a replica? Once there is no longer an available replica, we break out of the loop.